### PR TITLE
new postprocessor to compute spherical components of velocity field

### PR DESCRIPTION
**Pull request by cedrict**
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

This feature was missing so we implemented it. 

### Before your first pull request:

* [X ] I have read the guidelines in our [CONTRIBUTING.md](../CONTRIBUTING.md) document.

### For all pull requests:

* [ X] I have followed the [instructions for indenting my code](../CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [X ] I have tested my new feature locally to ensure it is correct.
* [ NA] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

File: source/postprocess/visualization/spherical_velocity_components.cc [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+// explicit instantiations
+namespace aspect
+{
+  namespace Postprocess
+  {
+    namespace VisualizationPostprocessors
+    {
+      ASPECT_REGISTER_VISUALIZATION_POSTPROCESSOR(SphericalVelocityComponents,
+                                                  "spherical velocity components",
+                                                  "A visualization output object that outputs the polar coordinates components"
```
Comment (tjhei) [inline source/postprocess/visualization/spherical_velocity_components.cc]: missing space at the end of each line

File: source/postprocess/visualization/spherical_velocity_components.cc [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+      template <int dim>
+      std::vector<std::string>
+      SphericalVelocityComponents<dim>::get_names () const
+      {
+        std::vector<std::string> names;
+        switch (dim)
+          {
+            case 2:
+              names.emplace_back("v_r");
```
Comment (tjhei) [inline source/postprocess/visualization/spherical_velocity_components.cc]: remove the switch and write it as:
```
names.emplace_back("v_r");
names.emplace_back("v_phi");
if (dim==3)
   names.emplace_back("v_theta");
```

**Comment by bangerth** [comment]
Out of curiosity, do you use anything other than just the vertical velocity? It's hard to imagine plotting the horizontal components given in phi/theta. But I could imagine visualizing the horizontal components as *vectors*.

File: source/postprocess/visualization/spherical_velocity_components.cc [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+                            std::vector<Vector<double> > &computed_quantities) const
+      {
+        const unsigned int n_quadrature_points = input_data.evaluation_points.size();
+        Assert (computed_quantities.size() == n_quadrature_points,    ExcInternalError());
+        Assert (computed_quantities[0].size() == dim,    ExcInternalError());
+
+        std::vector<Tensor<1,dim> > velocity(n_quadrature_points);
+        for (unsigned int q=0; q<n_quadrature_points; ++q)
+          {
+          for (unsigned int d = 0; d < dim; ++d)
```
Comment (bangerth) [inline source/postprocess/visualization/spherical_velocity_components.cc]: I think you didn't run the indentation script.

File: source/postprocess/visualization/spherical_velocity_components.cc [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            if (dim==2)
+               {
+               computed_quantities[q](0) =  std::cos(scoord[1])*velocity[q][0]+std::sin(scoord[1])*velocity[q][1]; // v_r
+               computed_quantities[q](1) = -std::sin(scoord[1])*velocity[q][0]+std::cos(scoord[1])*velocity[q][1]; // v_scoord[1]
+               }
+
+            if (dim==3)
+               { 
+               computed_quantities[q](0) = velocity[q][0]*(sin(scoord[2])*cos(scoord[1]))
+                                         + velocity[q][1]*(sin(scoord[2])*sin(scoord[1]))
```
Comment (bangerth) [inline source/postprocess/visualization/spherical_velocity_components.cc]: You use `std::sin` above, prepend the calls here with `std::` as well.

File: source/postprocess/visualization/spherical_velocity_components.cc (hunk 1 above)
Comment (cedrict) [reply source/postprocess/visualization/spherical_velocity_components.cc]: done.

File: source/postprocess/visualization/spherical_velocity_components.cc (hunk 2 above)
Comment (cedrict) [reply source/postprocess/visualization/spherical_velocity_components.cc]: done.

File: source/postprocess/visualization/spherical_velocity_components.cc (hunk 4 above)
Comment (cedrict) [reply source/postprocess/visualization/spherical_velocity_components.cc]: done.

File: source/postprocess/visualization/spherical_velocity_components.cc (hunk 3 above)
Comment (cedrict) [reply source/postprocess/visualization/spherical_velocity_components.cc]: indeed. done now.

File: source/postprocess/visualization/spherical_velocity_components.cc [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+      void
+      SphericalVelocityComponents<dim>::
+      evaluate_vector_field(const DataPostprocessorInputs::Vector<dim> &input_data,
+                            std::vector<Vector<double> > &computed_quantities) const
+      {
+        const unsigned int n_quadrature_points = input_data.evaluation_points.size();
+        Assert (computed_quantities.size() == n_quadrature_points,    ExcInternalError());
+        Assert (computed_quantities[0].size() == dim,    ExcInternalError());
+
+        std::vector<Tensor<1,dim> > velocity(n_quadrature_points);
```
Comment (tjhei) [inline source/postprocess/visualization/spherical_velocity_components.cc]:
```suggestion
        Tensor<1,dim> velocity;
```

File: source/postprocess/visualization/spherical_velocity_components.cc [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        for (unsigned int q=0; q<n_quadrature_points; ++q)
+          {
+            for (unsigned int d = 0; d < dim; ++d)
+              {
+                velocity[q][d] = input_data.solution_values[q][this->introspection().component_indices.velocities[d]];
```
Comment (tjhei) [inline source/postprocess/visualization/spherical_velocity_components.cc]:
```suggestion
                velocity[d] = input_data.solution_values[q][this->introspection().component_indices.velocities[d]];
```

File: source/postprocess/visualization/spherical_velocity_components.cc [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+                std::array<double,dim> scoord = aspect::Utilities::Coordinates::cartesian_to_spherical_coordinates(input_data.evaluation_points[q]);
+
+                if (dim==2)
+                  {
+                    computed_quantities[q](0) =  std::cos(scoord[1])*velocity[q][0]+std::sin(scoord[1])*velocity[q][1]; // v_r
```
Comment (tjhei) [inline source/postprocess/visualization/spherical_velocity_components.cc]: replace ``velocity[q]`` by ``velocity`` here (and everywhere below)

File: source/postprocess/visualization/spherical_velocity_components.cc (hunk 5 above)
Comment (cedrict) [reply source/postprocess/visualization/spherical_velocity_components.cc]: done.

File: source/postprocess/visualization/spherical_velocity_components.cc (hunk 6 above)
Comment (cedrict) [reply source/postprocess/visualization/spherical_velocity_components.cc]: done.

File: source/postprocess/visualization/spherical_velocity_components.cc (hunk 7 above)
Comment (cedrict) [reply source/postprocess/visualization/spherical_velocity_components.cc]: done.

File: source/postprocess/visualization/spherical_velocity_components.cc [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        Assert (computed_quantities.size() == n_quadrature_points,    ExcInternalError());
+        Assert (computed_quantities[0].size() == dim,    ExcInternalError());
+
+        Tensor<1,dim> velocity;
+        for (unsigned int q=0; q<n_quadrature_points; ++q)
+          {
+            for (unsigned int d = 0; d < dim; ++d)
+              {
+                velocity[d] = input_data.solution_values[q][this->introspection().component_indices.velocities[d]];
+
```
Comment (tjhei) [inline source/postprocess/visualization/spherical_velocity_components.cc]: I think you need to close the inner-most for loop here.

Comment (cedrict) [reply source/postprocess/visualization/spherical_velocity_components.cc]: hopefully got it right :)

Comment (tjhei) [reply source/postprocess/visualization/spherical_velocity_components.cc]: You know that you are forcing me to ask you to write a unit test now, right? :-)

Comment (cedrict) [reply source/postprocess/visualization/spherical_velocity_components.cc]: I need to write a test for another pull request. Can't you merge this and we'll get started on the tests ? :)

**Comment by tjhei** [comment]
I will add a test...

**Pull request merged**